[Code Together] Add support for qml.ResetError in default.mixed

wongwsvincent opened the pull request:
**Context:** Add support for qml.ResetError for default.mixed device, and a test of the gradient computation

**Description of the Change:** Added ResetError to the list of supported operations, and a test function for gradient check 

**Benefits:** Support single-qubit reset error channel that simulates the reset of qubits to 0 or 1 given a probability.

**Possible Drawbacks:**

**Related GitHub Issues:** [1530](https://github.com/PennyLaneAI/pennylane/issues/1530)

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -121,6 +121,9 @@
   an owner of the pruned tensor and its constituent observables, but leaves the
   original tensor in the queue without an owner.
 
+* The `qml.ResetError` is now supported for `default.mixed` device. 
```
Nice! :slightly_smiling_face:  Could you add yourself to the list of contributors too on line 137?

wongwsvincent replied:
Yep! I have just added my name.

antalszava commented on tests/ops/test_channel_ops.py:
```diff
@@ -279,6 +279,40 @@ def test_p0_p1_sum_not_normalized(self):
 
+    @pytest.mark.parametrize("angle", np.linspace(0, 2 * np.pi, 7))
+    def test_grad_reseterror(self, angle, tol):
+        """Test that analytical gradient is computed correctly for different states. Channel
```
Maybe worth adjusting the docstring here, as we're falling back to finite differences. `qml.ResetError` doesn't support analytic gradients at the moment, [`grad_method = "F"` here denotes this fact](https://github.com/PennyLaneAI/pennylane/blob/62f7c10766816322862d77d330ba83274fb1ff59/pennylane/ops/channel.py#L345).

wongwsvincent replied:
I have removed the word `analytical`. I hope it looks good now

antalszava commented on pennylane/devices/default_mixed.py:
```diff
@@ -91,6 +91,7 @@ class DefaultMixed(QubitDevice):
         "DepolarizingChannel",
         "BitFlip",
         "PhaseFlip",
+        "ResetError",
```
Awesome catch! :1st_place_medal:

antalszava reviewed (changes requested):
Hi @wongwsvincent, thank you for your contribution! :slightly_smiling_face: :tada: This is overall in great shape already, leaving some suggestions to refine it.

What would additionally remain, is adding tests for [applying the channel](https://github.com/PennyLaneAI/pennylane/blob/0088704258018171ad955d83238fb08ef94ed88d/tests/devices/test_default_mixed.py#L266) using `default.mixed`. The gradient test is already nice to have, these additional tests would just add an extra layer of unit testing.

wongwsvincent commented:
Hi @antalszava , I have made the changes as requested. This PR is ready for another review.

antalszava commented:
Hi @wongwsvincent, thanks! It seems that adding tests for [applying the channel](https://github.com/PennyLaneAI/pennylane/blob/0088704258018171ad955d83238fb08ef94ed88d/tests/devices/test_default_mixed.py#L266) using `default.mixed` would still remain.

Also, there seems to be a conflict in the changelog. This can be resolved by pulling master (`git checkout master` followed by `git pull`), switching to the feature branch (`git checkout add_reseterror_support`) and then merging `master` (`git merge master`). The culprit seems to be the changelog file. We'd like to have all the entries kept there, for some reason `git` could not resolve the addition there well.

wongwsvincent commented:
Hi @antalszava , thanks for reviewing and catching the missing tests! I have added a few tests for ```qml.ResetError``` in ```tests/devices/test_default_mixed.py```.  I have also fixed the conflicts in CHANGELOG.md. 

All checks look good without complaints. It should be ready for review.

antalszava commented on tests/templates/test_layers/test_strongly_entangling.py:
```diff
@@ -160,13 +160,6 @@ def circuit(weights, ranges=None):
             weights = np.random.randn(1, 2, 3)
             circuit(weights, ranges=[0])
 
-    def test_id(self):
```
Hi @wongwsvincent how come this test was removed?

wongwsvincent replied:
ops, sorry! I must have mistakenly removed it when I was resolving my merge conflicts. The test is added back.

antalszava commented on tests/ops/test_channel_ops.py:
```diff
@@ -279,6 +279,40 @@ def test_p0_p1_sum_not_normalized(self):
             channel.ResetError(1.0, 1.0, wires=0).kraus_matrices
 
+    @pytest.mark.parametrize("angle", np.linspace(0, 2 * np.pi, 7))
+    def test_grad_reseterror(self, angle, tol):
```
```suggestion
    def test_grad_reset_error(self, angle, tol):
```

antalszava commented on tests/devices/test_default_mixed.py:
```diff
@@ -99,11 +108,18 @@ def test_state_after_twoqubit(self, nr_wires, op, tol):
         assert np.allclose(dev.state, current_state, atol=tol, rtol=0)
 
-    @pytest.mark.parametrize("op", [AmplitudeDamping, DepolarizingChannel])
+    @pytest.mark.parametrize(
```
Nice adjustment of the tests here! :100:

wongwsvincent replied:
Thanks! =D

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -129,8 +129,10 @@
 * Create a separate requirements file for the CI issue , to have a separate requirements.txt (pinned)
 and requirements-ci.txt (unpinned). This latter would be used by the CI.
-  [(#1535)](https://github.com/PennyLaneAI/pennylane/pull/1535)
+  [(#1535)](https://github.com/PennyLaneAI/pennylane/pull/1535
```
```suggestion
  [(#1535)](https://github.com/PennyLaneAI/pennylane/pull/1535)
```

antalszava reviewed (commented):
Hi @wongwsvincent, the additional tests are looking awesome, thank you! :100: There seem to be small things to be adjusted. Overall, however, this is near approval.

wongwsvincent commented:
> Hi @wongwsvincent, the additional tests are looking awesome, thank you! 💯 There seem to be small things to be adjusted. Overall, however, this is near approval.

Thanks for the suggestions @antalszava ! All the issues are addressed. Ready for a (hopefully) final review.

antalszava reviewed (approved):
Looks good to me! :tada: Great one @wongwsvincent, thank you for your contribution! :slightly_smiling_face:
